### PR TITLE
add dotsnyk pattern matching

### PR DESCRIPTION
**Pull request by dekelund**
- [ ] Ready for review
- [ ] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

### CODE REVIEW

File: lib/scan.ts [hunk 1]
```diff
@@ -136,3 +140,13 @@ export async function scan(options: Options): Promise<PluginResponse> {
...
   }
 }
+
+function getExcludePatterns(projectRoot: string, dotSnykPath: string) {
+  try {
+    const dotSnyk: DotSnykConfig = parseDotSnyk(dotSnykPath);
+    return [dotSnykPath, ...parseUnmanagedExcludes(projectRoot, dotSnyk)];
+  } catch (err) {
+    // TODO(Tundra): Inform user that we failed to parse .snyk
+  }
```
Comment (dekelund) [inline lib/scan.ts:150]: maybe we should add an empty row between 150 and 151

File: lib/utils/dotsnyk/index.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+//     - “**/*.spec.ts” # files ending with “.spec.ts” from any directory
+//     - tests?/* # files directly inside “test” and/or “tests” directories
+//     - tests/** # all files and directories inside “tests”
+//
+
+export type Config = { exclude: { unmanaged: string[] } };
+
+export function parseDotSnyk(filePath: string): Config {
+  try {
+    const dotSnykFile = fs.readFileSync(filePath, 'utf-8');
```
Comment (dekelund) [inline lib/utils/dotsnyk/index.ts:22]: maybe rename dotSnykFile to content, bytes or similar. It's pretty obvious what it contains, due to the name of the function (`parseDotSnyk`).

File: package.json [hunk 3]
```diff
@@ -38,7 +39,7 @@
     "p-map": "^4.0.0",
     "tar": "^6.1.11",
     "tslib": "^2.0.0",
-    "uuid": "^8.3.2"
+    "yaml": "^1.10.2"
```
Comment (dekelund) [inline package.json:42]: oh it looks like we by accident removed `uuid`

File: lib/scan.ts [hunk 4]
```diff
@@ -136,3 +140,13 @@ export async function scan(options: Options): Promise<PluginResponse> {
     throw new Error(`Could not scan C/C++ project: ${error}`);
   }
 }
+
+function getExcludePatterns(projectRoot: string, dotSnykPath: string) {
```
Comment (dekelund) [inline lib/scan.ts:144]: should we create a test for this function?
(To test the integration between parseDotSnyk and parseUnmanagedExclude ?)

File: lib/scan.ts (hunk 1 above)
Comment (ekbsnyk) [reply lib/scan.ts:150]: like it 👍

File: lib/scan.ts (hunk 4 above)
Comment (ekbsnyk) [reply lib/scan.ts:144]: Yep, I think so.

File: lib/utils/dotsnyk/index.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    throw err;
+  }
+}
+
+export function parseUnmanagedExcludes(base: string, config: Config): string[] {
+  return toAbsolutePaths(base, config.exclude.unmanaged);
+}
+
+export function toAbsolutePaths(
```
Comment (ekbsnyk) [inline lib/utils/dotsnyk/index.ts:41]: For completeness + peace of mind maybe we should test this too?

File: lib/find.ts [hunk 6]
```diff
@@ -13,7 +14,7 @@ interface FileHandler {
 
 export async function find(
   src: string,
-  excluded: readonly Path[] = [],
+  excludePatterns: readonly string[] = [],
```
Comment (ekbsnyk) [inline lib/find.ts:17]: Maybe we should test that find() behaves properly when excludePatterns is empty, or in other possible edge cases?

File: lib/scan.ts [hunk 7]
```diff
@@ -136,3 +140,13 @@ export async function scan(options: Options): Promise<PluginResponse> {
...
+  try {
+    const dotSnyk: DotSnykConfig = parseDotSnyk(dotSnykPath);
+    return [dotSnykPath, ...parseUnmanagedExcludes(projectRoot, dotSnyk)];
+  } catch (err) {
+    // TODO(Tundra): Inform user that we failed to parse .snyk
```
Comment (ekbsnyk) [inline lib/scan.ts:149]: I guess we should implement error handling here before merging

File: lib/utils/dotsnyk/index.ts (hunk 2 above)
Comment (ekbsnyk) [reply lib/utils/dotsnyk/index.ts:22]: fileContents?

File: test/dotsnyk.test.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import { parseDotSnyk, parseUnmanagedExcludes } from '../lib/utils/dotsnyk';
+import { join } from 'path';
+
+describe('parseDotSnyk', () => {
```
Comment (ekbsnyk) [inline test/dotsnyk.test.ts:4]: Would it add any value to have a test where we read from an actual .snyk file to approximate how it's used in real life?

File: lib/utils/dotsnyk/index.ts (hunk 5 above)
Comment (dekelund) [reply lib/utils/dotsnyk/index.ts:41]: Yep, this is copy-pasted from another file (used by scan.ts), and we use a copy to be able to extract dotsnyk/index.ts later to a separate lib.

So we could copy following test more or less: https://github.com/snyk/snyk-cpp-plugin/blob/master/test/scan.test.ts#L59

File: lib/utils/dotsnyk/index.ts (hunk 2 above)
Comment (dekelund) [reply lib/utils/dotsnyk/index.ts:22]: sounds good, maybe scratch the plural version and just call it fileContent?

File: lib/scan.ts (hunk 4 above)
Comment (dekelund) [reply lib/scan.ts:144]: 👍

File: lib/find.ts (hunk 6 above)
Comment (dekelund) [reply lib/find.ts:17]: (I just mean old tests, more than a week ago, calls find without second argument)

File: lib/find.ts [hunk 9]
```diff
@@ -31,7 +32,9 @@ export async function find(
     fileResults.push(path);
   };
 
-  await traverse(src, handler, (p) => excluded.includes(p));
+  await traverse(src, handler, (path) => {
+    return !!excludePatterns.find((pattern) => minimatch(path, pattern));
```
[the remark this answers is not in the document]
Comment (dekelund) [reply lib/find.ts:36]: yea maybe. In our case it's because it provide us with the small subset that we only need.
(If we used the snyk fast-glob it would touch the filesystem which we already have code to do, that does it pretty efficient. Otherwise we need I/O calls times two)

File: lib/scan.ts (hunk 1 above)
Comment (dekelund) [reply lib/scan.ts:150]: lets make it happen :)

File: package.json (hunk 3 above)
Comment (ekbsnyk) [reply package.json:42]: have added it back in, will commit next 👍

**Review by danlucian — APPROVED** [review]
